src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionLiteral.cs

```diff
@@ -0,0 +1,121 @@
+            // PROTOTYPE: Should [] be emitted as Array.Empty<T>()?
+            var initialization = (arrayLength == 0)
+                ? null
+                : new BoundArrayInitialization(
+                    syntax,
+                    isInferred: false,
+                    initializers.SelectAsArray(e => VisitExpression(e)));
+            var array = new BoundArrayCreation(
+                syntax,
+                ImmutableArray.Create<BoundExpression>(
+                    new BoundLiteral(
+                        syntax,
+                        ConstantValue.Create(arrayLength),
+                        _compilation.GetSpecialType(SpecialType.System_Int32))),
+                initialization,
+                arrayType)
+            { WasCompilerGenerated = true };
+
+            if (spanConstructor is null)
+            {
+                return array;
+            }
+
+            Debug.Assert(TypeSymbol.Equals(array.Type, spanConstructor.Parameters[0].Type, TypeCompareKind.AllIgnoreOptions));
+            return new BoundObjectCreationExpression(syntax, spanConstructor, array);
+        }
+
+        public override BoundNode? VisitCollectionInitializerCollectionLiteralExpression(BoundCollectionInitializerCollectionLiteralExpression node)
+        {
+            Debug.Assert(!_inExpressionLambda);
+            Debug.Assert(node.Type is { });
+
+            var syntax = node.Syntax;
+            var collectionType = node.Type;
+
+            BoundExpression rewrittenReceiver;
+            if (collectionType is TypeParameterSymbol typeParameter)
+            {
+                // PROTOTYPE: If we support _inExpressionLambda, see VisitNewT()
+                // which does not call MakeNewT() in that case.
+                rewrittenReceiver = MakeNewT(syntax, typeParameter);
+            }
+            else
+            {
+                var collectionCreation = node.CollectionCreation;
+                Debug.Assert(collectionCreation is { });
+                rewrittenReceiver = VisitExpression(collectionCreation);
+            }
+
+            // Create a temp for the collection.
+            BoundAssignmentOperator assignmentToTemp;
+            BoundLocal temp = _factory.StoreToTemp(rewrittenReceiver, out assignmentToTemp, isKnownToReferToTempIfReferenceType: true);
+            var initializers = node.Initializers;
+            var sideEffects = ArrayBuilder<BoundExpression>.GetInstance(initializers.Length + 1);
+            sideEffects.Add(assignmentToTemp);
+
+            AddPlaceholderReplacement(node.Placeholder, temp);
+
+            foreach (var initializer in initializers)
+            {
+                BoundExpression? rewrittenInitializer;
+                switch (initializer)
+                {
+                    case BoundCollectionElementInitializer element:
+                        rewrittenInitializer = MakeCollectionInitializer(temp, element);
+                        break;
+                    default:
```

Comment:
feels like this would read just a little more clearly as `if (initializer is not BoundCollectionElementInitializer element) { throw; } rewrittenInitializer = MakeCollectionInitializer(temp, element);`. Unless is this is planned to be expanded later?